[Feature] ADD: show create routine load

stalary opened the pull request:
## Proposed changes

Add show create routine load

Fixed #5645

stalary committed:
ADD: show create routine load

stalary committed:
ADD: show create routine load

stalary committed:
ADD: show create routine load

stalary committed:
ADD: show create routine load

stalary committed:
Merge branch 'show_create_routine_load' of https://github.com/stalary/incubator-doris into show_create_routine_load

stalary commented:
https://github.com/apache/incubator-doris/issues/5645

stalary committed:
FIX: fix some bug

stalary committed:
FIX: fix some bug

stalary committed:
ADD: add end_line

stalary committed:
ADD: add doc

stalary committed:
MOD: 修改sidebar

morningman committed:
[Enhance] Support show unrecoverable tablets (#6045)

* [Enhance] Support show unrecoverable tablets

The unrecoverable tablets are tablets which non of their replicas are healthy.
We should be able to find out these tablets then manual intervention.

And these tablets should not be added to the tablet scheduler.

e0c9 committed:
[Feature] Support recording custom number of backup and restore task information (#5947)

* Record all backup jobs and support where clause

Commit:
fix spring boot web maximum upload file limit config (#6070)

Co-authored-by: zouxinyi <zouxinyi@baidu.com>

morningman committed:
[Build][ARM] Fix some compilation problems on ARM64 (#6076)

1. Disable libhdfs3 on ARM, because it doesn't support ARM now.
2. Add compilation doc for ARM64

EmmyMiao87 commented:
In fact, these information will be displayed in the results of show routine load.
Under what circumstances will a show create routine load syntax be needed?

morningman commented:
> In fact, these information will be displayed in the results of show routine load.
> Under what circumstances will a show create routine load syntax be needed?

Just like `show create table`, user may may need to create routine load job again.

morningman committed:
[Bug] Fix bug that select into outfile in parquet format may cause NPE (#6054)

1. check the parquet schema property on FE side.
2. auto generate parquet schema if not specified.

stalary commented:
For example, I stopped it and then tried to restart it

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/ShowCreateRoutineLoadStmt.java:
```diff
@@ -0,0 +1,73 @@
+    
+    @Override
+    public void analyze(Analyzer analyzer) throws AnalysisException {
+        if (labelName == null) {
```
Label_name in the grammar file is not optional, so it will not be empty.

stalary replied:
OK, thanks for the warning

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/qe/ShowExecutor.java:
```diff
@@ -1834,7 +1837,51 @@ private void handleShowQueryProfile() throws AnalysisException {
+        // if job exists
+        RoutineLoadJob routineLoadJob;
+        try {
+            routineLoadJob = Catalog.getCurrentCatalog().getRoutineLoadManager()
```
getRoutineLoadManager().checkPrivAndGetJob()

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/qe/ShowExecutor.java:
```diff
@@ -1834,7 +1837,51 @@ private void handleShowQueryProfile() throws AnalysisException {
+        RoutineLoadJob routineLoadJob;
+        try {
+            routineLoadJob = Catalog.getCurrentCatalog().getRoutineLoadManager()
+                    .getJob(showCreateRoutineLoadStmt.getDb(),
```
getJob() function will not displayed the routine load in the stopped state.

EmmyMiao87 commented:
You need to consider a question. 
Whether the routine load after stopping also needs to be displayed.

The routine load after stopping may have the same name. 
So if show, it may produce multiple results.
The one way is that using 'show create routine load' to display running routine load, using 'show all create routine load' to displays all routine loads with the same name include stopped.

EmmyMiao87 commented:
Fixed #5645

stalary commented:
> You need to consider a question.
> Whether the routine load after stopping also needs to be displayed.
> 
> The routine load after stopping may have the same name.
> So if show, it may produce multiple results.
> The one way is that using 'show create routine load' to display running routine load, using 'show all create routine load' to displays all routine loads with the same name include stopped.
At the moment my requirement is just to show running, do we need to support all of them？Show all create routine load is not very useful to me right now.

stalary commented:
> You need to consider a question.
> Whether the routine load after stopping also needs to be displayed.
> 
> The routine load after stopping may have the same name.
> So if show, it may produce multiple results.
> The one way is that using 'show create routine load' to display running routine load, using 'show all create routine load' to displays all routine loads with the same name include stopped.

Well, I've decided to support both

EmmyMiao87 commented:
> At the moment my requirement is just to show running, do we need to support all of them？Show all create routine load is not very useful to me right now.

This is no problem. 
It needs to be stated on the document that only the routine load of running or paused is displayed.

EmmyMiao87 committed:
[Docs] Fix README.md (#6084)

e0c9 committed:
fix (#6081)

stalary committed:
MOD: support all

stalary commented:
PTAL @EmmyMiao87

HappenLee committed:
SIMD instruction speed up the storage layer (#6089)

* SIMD instruction speed up the storage layer

* 1. add DECHECK in power of 2 int32
2. change vector to array deduce the cost

stdpain committed:
fix stdfs link error in some gcc version (#6090)

EmmyMiao87 commented on docs/zh-CN/sql-reference/sql-statements/Data Manipulation/SHOW CREATE ROUTINE LOAD.md:
```diff
@@ -0,0 +1,42 @@
+
+# SHOW CREATE ROUTINE LOAD
+## description
+    该语句用于展示用户自定义例行导入作业的创建语句
```
```suggestion
    该语句用于展示例行导入作业的创建语句
```

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/load/routineload/KafkaRoutineLoadJob.java:
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.doris.load.routineload;
 
+import java.util.HashMap;
```
Pay attention to the ordering of imports

stalary replied:
ok, I'll adjust it later

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/qe/ShowExecutor.java:
```diff
@@ -1834,7 +1837,56 @@ private void handleShowQueryProfile() throws AnalysisException {
+        if (showCreateRoutineLoadStmt.isIncludeHistory()) {
+            List<RoutineLoadJob> routineLoadJobList = new ArrayList<>();
+            try {
+                routineLoadJobList = Catalog.getCurrentCatalog().getRoutineLoadManager().getJob(dbName, labelName, true);
```
The 'getJob' function will return null when there is no routine load has been found.

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/qe/ShowExecutor.java:
```diff
@@ -1834,7 +1837,56 @@ private void handleShowQueryProfile() throws AnalysisException {
+                        dbName,
+                        tableName,
+                        PrivPredicate.LOAD)) {
+                    ErrorReport.reportAnalysisException(ErrorCode.ERR_TABLEACCESS_DENIED_ERROR, "LOAD",
```
If user don’t have permission, we can simply not display it without reporting an error

EmmyMiao87 commented on fe/fe-core/src/main/java/org/apache/doris/qe/ShowExecutor.java:
```diff
@@ -1834,7 +1837,56 @@ private void handleShowQueryProfile() throws AnalysisException {
+                routineLoadJob = Catalog.getCurrentCatalog().getRoutineLoadManager().checkPrivAndGetJob(dbName, labelName);
+                // get routine load info
+                rows.add(Lists.newArrayList(String.valueOf(routineLoadJob.getId()), showCreateRoutineLoadStmt.getLabel(), routineLoadJob.getShowCreateInfo()));
+            } catch (Exception e) {
```
The exception caught here is too broad

EmmyMiao87 commented:
Please add some UT~

stalary committed:
MOD: mod some code
